### PR TITLE
Forward-port assorted upgrade checks 2.x to 3.x

### PR DESCRIPTION
**Pull request by dylant-da**
The following commits are cherry-picks:

```
[Compiler] types from the non stable part of stdlib are unserialiazable (#19345)
Port key upgrade static checks from 2.9 to 2.x (#19565)
Pick "Ignore "utility packages" in upgrade checks (#19392)"
Port instance tests (pick some parts of "Backport Dylan's interface upgrade checks (#19380)")
Pick "Force that upgrading types must belong to identically-named packages (#19357)"
Pick "ignore non serializable types during upgrade validation (#19356)"
Cherrypick "upgrades check all packages (#19328)", dupe ValidUpgrade
Forwardport daml#19392 and canton#19524
```

### CODE REVIEW

File: sdk/daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 1]
```diff
@@ -1057,47 +1057,7 @@ private[archive] class DecodeV1(minor: LV.Minor) {
...
-    val exerciseByKey = v1_11
-    val internedTypes = v1_11
-    val choiceObservers = v1_11
-    val bigNumeric = v1_13
-    val exceptions = v1_14
-    val basicInterfaces = v1_15
-    val choiceFuncs = v1_dev
-    val choiceAuthority = v1_dev
-    val natTypeErasure = v1_dev
-    val packageUpgrades = v1_dev
```
Comment (samuel-williams-da) [inline sdk/daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:1084]: Should this not have been v1_16? or did v16 not make it to `main`?

File: sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala [hunk 2]
```diff
@@ -1283,7 +1283,6 @@ object Ast {
         metadata: PackageMetadata,
     ): GenPackage[E] = {
       val isUtilityPackage =
-        metadata.name == "daml-prim" || metadata.name == "daml-stdlib" ||
```
Comment (samuel-williams-da) [inline sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala]: Oh, so prim and stdlib don't export anything serializable anymore? Did we update them on the 3x branch?

**Review by samuel-williams-da — APPROVED** [review]
I've only reviewed code that isn't direct cherry-picks. If this is acceptable, looks good to me.

File: sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala (hunk 2 above)
Comment (samuel-williams-da) [reply sdk/daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala]: I noticed later on you updated RoundingMode to not be serializable, so I guess the answer to this is yes.

**Comment by samuel-williams-da** [comment]
I see you're getting `Script.FailedCmd unexpected cause: Boxed Exception\n\n` in CI, those boxed exceptions aren't great. you could be able to find the serializer for FailedCmd and add some cases to pull them out. They come from `ExecutionException`, which should let you see the cause.

**Comment by dylant-da** [comment]
Looks like it was because I got the comparison for `supportsPersistedPackageVersion` the wrong way around, which was causing the assertion on line 98 of `daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala` to fail:

```
assert(packageVersion.isEmpty || version >= TransactionVersion.minPackageVersion)
```

I've fixed that now - hopefully this causes them to pass

**Review by samuel-williams-da — APPROVED** [review]
I assume the package id changes were expected? What is different about them?

**Comment by dylant-da** [comment]
The differences come from adding the serializable tag to CallStack and SrcLoc

**Comment by samuel-williams-da** [comment]
Ah okay, thats fine then

**Pull request merged**